[Jung] 미션4 모든 기물 배치하기 PR입니다

rla36 opened the pull request:
안녕하세요 Jung입니다.

- 미션 4 에서는 
utils.StringUtils클래스, Name enum이 추가되었고, 기존의 Pawn클래스가 Piece로 리네이밍 됐습니다.
Name 이넘에서는 체스 말들의 이름을 다루도록했고, 각 말들의 흑/백 representation을 생성자에 넣어 값으로 갖도록했습니다.
Board클래스에서는 initialize()를 통해 32개의 말을 체스판에 추가하도록했는데, Pawn을 initialize하는 메서드와 Pawn외 다른 말들을 initialize하는 메서드를 분리해 작성 후, initialize()메서드 안에 넣었습니다.
기존의 클래스, 메서드들을 최대한 재사용하는 방향으로 진행해서, 기존에 만들었지만 현재 미션에 필요하지 않은 메서드도 다량 포함되어 있다는 점 양해해주시길 바랍니다..

- 루카스의 미션4 진행하며 고민이 많았는데, 요구사항을 보면 다음과 같은 이야기가 있었습니다.

> Piece는 값(value) 객체여야 한다.

> 즉, private 생성자를 가져야 하며, 인스턴스를 생성한 이후에는 인스턴스의 상태 값을 변경할 수 없어야 한다.

검색해본 결과 "two value objects are equal when they have the same value, not necessarily being the same object." 라고 해서, 같은 객체가 아니더라도 값이 같다면 비교시(equals로?) true여야 한다는 말로 이해를 했습니다. 체스판 위에 white Pawn이 8개지만, 모두 흰색이고 name이 pawn으로 같으므로 8개 모두 같은 객체여야 한다는 뜻 같았습니다.
그래서 equals메서드를 오버라이딩해서 name, color, representation이 동일하다면 같은 piece라는 결과를 보이도록해야하나 고민했습니다. 다른 분들이 미션4 진행하신 것을 보니 그렇게 하신분은 없어서 제가 잘못이해한것 같아 실제로 구현하지는 않았습니다.

그렇다면, **private생성자를 가지고, 인스턴스 생성 후에도 인스턴스 상태값을 변경할 수 없다 -> 이 두 조건은 value object를 만들기 위한 필요조건이지만, 충분조건은 아닌것인가요?** 저 두 조건을 만족하도록 미션을 진행했음에도 equals를 오버라이딩 하지 않았기 때문에 createWhitePawn()메서드로 만들어진 두 인스턴스가 equals 비교시 false가 나왔습니다. (아무리 생각해도 제가 무언가를 잘못하고 있는 것같아 커밋에는 포함하지 않았습니다!!) 

리뷰해주셔서 감사합니다!

rla36 committed:
\n을 제거하기 위한 StringUtils클래스 생성

운영체제별로 다른 newLine 문자를 제거하기 위해
System.getProperty("line.separator")로 대체했습니다.
매개변수로 String을 받아 새로운 라인을 추가하고 리턴하는
appendNewLine메서드를 구현했습니다.

rla36 committed:
appendNewLine()메서드를 사용해 \n 제거

기존에 \n을 사용해 다음 라인으로 넘어갔던 부분을 제거하고
StringUtils클래스에 정의된 appendNewLine()메서드를 사용하도록 했습니다.

rla36 committed:
test: \n 제거하고 appendNewLine으로 대체

기존에 print()메서드로 리턴되는 스트링 값이 출력 기댓값과 일치하는지
확인하는 테스트 메서드에서, 기대스트링에  \n 을 사용했기 때문에 System.getProperty("line.separator")과 일치하지 않아 테스트가 통과되지 않았습니다.
기대스트링도 System.getProperty("line.separator")을 사용하도록
수정했습니다.

rla36 committed:
Name enum추가

Pawn외 말들 관리하기 위해 Name enum을 추가했습니다.
각 말들의 이름과 함께 representation도 enum의 멤버변수로 갖도록
했습니다.

rla36 committed:
Pawn 클래스 Piece로 이름변경

Pawn클래스를 Piece로 이름 변경 후 생성자에 Color과 Name을 받도록
했습니다. 외부에서 생성자를 사용할수없도록 private생성자로 만들었고,
대신 새로운 인스턴스를 리턴하는 public메서드를 만들어 외부에서
인스턴스를 생성하도록했습니다.

rla36 committed:
Pawn을 Piece로 교체한 후  Board와 테스트 클래스 수정

삭제된 Pawn기본생성자, 외부에서 Piece 인스턴스 생성 불가한 점, 등을
고려해 기존 test들이 통과하도록 수정

rla36 committed:
Knight representation n/N으로 수정

rla36 committed:
인스턴스변수에 final 접근제어자 설정

변경못하도록 final 설정

rla36 committed:
test : 말 생성 팩토리 메서드 검증 테스트 추가

createWhitePawn() 등으로 생성되는 말의 색, 이름 representation이
정확한지 검증하는 테스트를 작성했습니다.

rla36 committed:
test: initialize() print() 검증 테스트 작성

initialize로 흑백의 말이 각각 16개씩 보드에 추가되는 것을 확인하는
테스트를 작성했습니다.

rla36 committed:
feat: initialize로 체스판에 32개 말 올리는 기능 추가

흑백 폰은 8개씩 추가해야하므로 중복이 발생해서 별도의
pawnInitialize메서드로 빼내었습니다. 이전단계에 작성했던  메서드를 최대한
유지하는 방향으로 진행해 getPieceResult  메서드로 각 흑/백 말 리스트에 있는
모든 말의 대표값을 스트링으로 리턴하도록했습니다. 8번재 말에서 새로운 라인을 추가하도록했습니다.

rla36 committed:
initialize 중복제거

createWhitePawn() createBlackQueen() 메서드를 사용하지 않고
createPiece(color, name)을 사용해 중복제거했습니다.

rla36 committed:
isBlack isWhite검증 테스트 작성

isBlack은 Piece의 색이 검정일때 true, 흰색일때 false
isWhite는 Piece의 색이 흰색일 때 true, 검정일 때 false
를 리턴하는지 검증하는 테스트를 작성했습니다.

rla36 committed:
feat: isBlack isWhite 메서드 추가

Piece의 색에 따라 isWhite() isBlack()이 true/false를 알맞게 리턴하는
메서드를 추가했습니다.

rla36 committed:
전반적인 리팩토링

유효하지 않은 테스트 삭제 및 메소드 리팩토링 진행했습니다.

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,114 @@
 package net.jung.chess;
 
-import net.jung.chess.pieces.Pawn;
+import net.jung.chess.pieces.Piece;
```
`pieces` 안에 Piece가 있는건 그냥 동어 반복의 느낌이네요!

rla36 replied:
pieces안에 Piece클래스만 있으니까 정말 그런것같네요 
Color, Type enum도 Piece의 속성이니까 pieces디렉토리에 이동하도록 수정했습니다!

rla36 replied:
미션 5 요구사항 보니 Color과 Type이넘도 Piece 내부에 포함시키라고 하더라고요
그래서 pieces디렉토리를 없애고 chess바로 아래에 Pieces가 들어가도록 수정하려고 합니다!

ksundong replied:
앗 되도록이면 요구사항을 미리 보지는 말아주세요. 😊
왜냐하면 요구사항에 맞게 변화시키면서 어떤 코드가 더 유연한지 깨달을 기회가 줄어들거든요!

rla36 replied:
넵넵 두번째 코멘트는 미션4 수정하고 일단 미션5로 넘어가고 나서 단 코멘트입니다 ㅎㅎ

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,114 @@
-        if(pawn.getColor() == Color.WHITE)
-        whitePawnList.add(pawn);
+    public void addWhitePiece(Piece piece) {
+        if (piece.getColor() == Color.WHITE)
+            whitePieceList.add(piece);
```
저는 되도록이면 중괄호(`{}`)는 안빼는 것을 추천드려요.

ksundong replied:
그리고 요구사항에 `isWhite()`같은 요구사항이 있던걸로 기억하는데, 그걸 활용해주시면 더 좋을 것 같아요!

rla36 replied:
넵 수정했습니다 새로 추가된 메서드가 어디 적용될 수 있을지 쓰임새를 생각해보는 습관을 들여야겠네용 감사합니당

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,114 @@
+    public void addPiece(Piece piece){
+        if(piece.getColor()==Color.WHITE) {
+            whitePieceList.add(piece);
+            return;
```
early return을 해준 것은 좋았는데, 굳이 메서드를 하나 더 만들어줬어야 하나 싶긴하네요!
의도가 명확한 메서드를 사용하도록 이것은 제거해주는 것도 생각해보세요!

rla36 replied:
아래 Board클래스의 68~75 줄 코멘트 주신 것에 답변했습니다!

> 이 메서드를 만든 이유가 중복제거를 하기 위해서였습니다.
whitePiece와 blackPiece가 위의 말들을 8개씩 갖고 있고, initialize를 하면서 8개의 말을 whitePieceList와 blackPieceList에 추가해야합니다.
> 그런데  createWhiteRook(), createWhiteKnight()...에서 createWhiteRook()까지 8개의 말을 만들어 addWhitePiece()를 해주고, 검은 말을 더할 떄 또 createBlackRook(), ... createBlackRook로 만들어서 addBlackPiece로 더해주면 비슷한 메서드들 8개가 2번 반복되므로 제거할 수 있는 중복이라고 생각을 했습니다.
> 그래서 색깔을 매개변수로 받아 white/black 각각의 Rook, Knight, ... Rook까지 8개를 만들고, addPiece메서드를 사용해 색깔을 판단한 후 적절한 List에 더해주면 될 것이라고 생각해 저렇게 작성했습니다.
> addWhitePiece()  addBlackPiece() 메서드가 있는데도 addPiece메서드를 만든 이유도 여기에서 중복제거를 하기 위해서였습니다. 그런데 의도가 명확한 메서드만을 남기는것이 좋다고하시니 , 어느 정도의 중복을 허용하는 방향으로 수정을 해야할 것같다고 생각했습니다!

중복되는 코드가 싫어서 다 제거하기 위해 팩토리메서드(createWhitePawn())을 쓰지 않고 (createPiece(Color.WHITE, Type.Pawn)으로 생성했는데, 아래 createPiece코드도 제거하라고 말씀하셨으니 어차피 다 새로짜야할것같네요ㅜ 
아무튼 코멘트 감사합니다! 수정하겠습니다

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,114 @@
 
     public int size() {
-        return whitePawnList.size() + blackPawnList.size();
+        return whitePieceList.size() + blackPieceList.size();
```
아래에 만들어준 `Size()` 메서드를 재활용해보는 것은 어떨까요?

rla36 replied:
디온 질문이 있는데요!! 
현재 size()메서드를 호출하면 1) whitePawnList, blackPawnList에 접근해 바로 2) size()를 리턴합니다.

그런데 만약 제가 만들었던 whitePieceListSize() 와  blackPieceListSize() 메서드를 사용하게 되면, 
  
    public int size() {         return whitePieceListSize() + blackPieceListSize();  }
    public int whitePieceListSize() {        return whitePieceList.size();    }
    public int blackPieceListSize() {        return blackPieceList.size();    }


.1) whitePieceListSize() blackPieceListSize()메소드를 호출해  2) 메서드 안의 whitePieceList로 가서 3) size를 리턴하게 됩니다

절차가 하나 늘어난다고 볼 수 있는데 효율성이 떨어지는것 아닌가요??
whitePieceListSize blackPieceListSize메서드를 구현한 것은 whitePieceList와 blackPieceList가 private하기 때문에 클래스 외부에서 말의 개수를 알고 싶을 때 이 public 메서드를 쓰도록한 것입니다. 
그래서 whitePieceList, blackPieceList와 같은 클래스 내에 있는 size()메서드를 쓸 때 꼭 만들었던 메서드를 사용하지 않고 List에 직접 접근해 쓰는 것이 낫지 않나 궁금합니다!!

ksundong replied:
@rla36 자 우리는 효율성에 대해서 그리 깊게 고민하지 않아도 됩니다.
우리는 우리가 짠 코드가 얼마나 유지보수 되기 쉬운지를 더 우선순위로 두는게 좋아요.
특히 메서드 콜이 많아진다고 이것이 엄청나게 느려지거나 하지 않습니다.

예를 들어, 만약 white piece의 size에 별도의 로직이 들어가는 경우, 현재 코드는 변경 포인트가 두 곳이 됩니다.
반면 메서드콜을 하는 경우는 변경 포인트가 단 한 곳만 존재하게됩니다.
메서드 콜의 효율성은 굳이 따질 필요가 없는 정도의 차이입니다!

rla36 replied:
아하 알겠습니다!! 이해 잘됐어용 감사합니다 ㅎㅎ

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,114 @@
+    }
+
+    public void pawnInitialize(Color color){
+        int MAX_PAWNS_SIZE = 8;
```
지역변수는 카멜 케이스, 상수는 스크리밍 스네이크 케이스를 지켜주세요~

rla36 replied:
static final 변수를 메서드 내부로 끌어온 후 수정을 안했나봐요 ㅜㅜ 고쳤습니다!!

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,114 @@
+        piecesExceptPawnInitialize(Color.WHITE);
+    }
+
+    public void pawnInitialize(Color color){
```
보시면 `addPiece`는 동사가 앞에서부터 시작하고, 이 메서드는 `pawn`이라는 명사가 앞에서부터 시작하죠?
저는 이것또한 일관성의 범주라고 생각합니다.

코드를 읽었을 때,
`board.pawnInitialize()`
`board.initializePawns()`
어떤 부분이 더 나은지 생각해봐주세요!

rla36 replied:
메서드 이름 지을 떄 동사를 많이 쓰니까 중요한 동사를 앞으로 빼내겠습니다!

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,114 @@
-    public String getWhitePawnsResult() {
-        return getPawnsResult(whitePawnList);
+    public void piecesExceptPawnInitialize(Color color){
+        addPiece(Piece.createPiece(color, Name.ROOK));
+        addPiece(Piece.createPiece(color, Name.KNIGHT));
+        addPiece(Piece.createPiece(color, Name.BISHOP));
+        addPiece(Piece.createPiece(color, Name.QUEEN));
+        addPiece(Piece.createPiece(color, Name.KING));
+        addPiece(Piece.createPiece(color, Name.BISHOP));
+        addPiece(Piece.createPiece(color, Name.KNIGHT));
+        addPiece(Piece.createPiece(color, Name.ROOK));
```
각 Piece별로 메서드를 만들어주는 선택지도 있었는데, 그렇게 하지 않으신 이유가 있나요?

rla36 replied:
이 메서드를 만든 이유가 중복제거를 하기 위해서였습니다.
whitePiece와 blackPiece가 위의 말들을 8개씩 갖고 있고, initialize를 하면서 8개의 말을 whitePieceList와 blackPieceList에 추가해야합니다. 
그런데 createWhiteRook(), createWhiteKnight()...에서 createWhiteRook()까지 8개의 말을 만들어 addWhitePiece()를 해주고, 
검은 말을 더할 떄 또 createBlackRook(), ... createBlackRook로 만들어서 addBlackPiece로 더해주면 비슷한 메서드들 8개가 2번 반복되므로 제거할 수 있는 중복이라고 생각을 했습니다.

그래서 색깔을 매개변수로 받아 white/black 각각의 Rook, Knight, ... Rook까지 8개를 만들고, addPiece메서드를 사용해 색깔을 판단한 후 적절한 List에 더해주면 될 것이라고 생각해 저렇게 작성했습니다.
addWhitePiece() addBlackPiece() 메서드가 있는데도 addPiece메서드를 만든 이유도 여기에서 중복제거를 하기 위해서였습니다. 
그런데 의도가 명확한 메서드만을 남기는것이 좋다고하시니 , 어느 정도의 중복을 허용하는 방향으로 수정을 해야할 것같다고 생각했습니다!

ksundong replied:
비슷하다와 중복은 조금 다른 개념입니다.
물론 중복되는 로직이 있긴 하겠지만, 코드의 가독성 또한 고려를 해봐야 할 사항이에요.
보통 도메인과 연관된 로직은 명시적으로 표현해주는 편이 좋고, 그 외의 경우에는 이렇게 공통되는 로직을 추출해도 좋아요.

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,114 @@
-        int colorPawnSize = pawnList.size();
-        for (int i = 0; i < colorPawnSize; i++) {
-            pawnReps.append(pawnList.get(i).getRepresentation());
+        for (int i = 0, colorPawnSize = pieceList.size(); i < colorPawnSize; i++) {
```
```suggestion
        for (int i = 0; i < pieceList.size(); i++) {
```

또는 사실 이 size가 결국 8이기 때문에 `MAX_PAWNS_SIZE`를 `MAX_PIECES_SIZE` 정도로 바꾸고 활용하는 방안도 있겠네요.

rla36 replied:
가독성은 조금 떨어지지만 for문 돌떄마다 size() 연산을 하는 것이 싫어 
for(int i = 0, colorPawnSize = pieceList.size(); i < colorPawnSize; i++)와 같이 작성했습니다.

size가 8이 아닐때도 고려해서 MAX_PIECES_SIZE를 쓰지 않고 size()를 넣었습니다!

ksundong replied:
Size가 8이 아닌 경우는 어떤 경우인가요?
그렇게 되는 경우는 애플리케이션이 오류를 내야하는 것이 정상적인 동작 아닐까요?
차라리 연산을 하지 않도록 위로 꺼내주시는 것도 방법입니다.

rla36 replied:
지난 단계에서 위로 꺼내는 방법을 썼었는데 간결하게 하라는 코멘트가 달려서 그에 따라 수정된 결과가 저것입니다ㅜㅜ 
size 8이 아닌 경우는 말이 죽었을 때가 있을 것 같은데 이번 미션에서는 그럴일은 없으니까 그냥 maxPiecesSize =16 으로  수정하거나 말씀하신 pieceList.size()를 그냥 쓰는 것이 좋을것같네여 후자로 하겠습니다!

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,114 @@
 
-    public String getPawnsResult(List<Pawn> pawnList) {
+    public String getPiecesResult(List<Piece> pieceList) {
         StringBuilder pawnReps = new StringBuilder();
```
`pawnReps`같이 줄이기보다는
`piecesRepresentation`으로 바꾸고, 메서드도 `getPiecesRepresentation()` 정도로 고치는게 일관적일것 같네요.

rla36 replied:
수정했습니다!

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,114 @@
         return pawnReps.toString();
     }
 
     public String print() {
```
`print()`라는 이름 말고 더 좋은 이름을 고민해봐주세요!

rla36 replied:
스트링을 리턴해주는 메서드이므로 boardLayoutToString으로 고쳐봤습니다!

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,114 @@
-    public void reset(){
-        whitePawnList.clear();
-        blackPawnList.clear();
+    public void reset() {
+        whitePieceList.clear();
+        blackPieceList.clear();
     }
```
오 `reset()`도 만들어주셨네요.

rla36 replied:
넵 사용자가 start여러번 누르면 그때마다 재시작하도록 했습니당

ksundong commented on src/main/java/net/jung/chess/Name.java:
```diff
@@ -0,0 +1,26 @@
+package net.jung.chess;
+
+public enum Name{
```
`Name`이 조금 아쉽네요 ㅠㅠ
더 적합한 표현을 생각해봐주세요!

rla36 replied:
체스말을 표현할 때 색/이름으로 한다고해서 단순히 Name으로 직역했는데 좀 부적절했던 것 같습니다..ㅋㅋㅋ
다른 분들 것 참고해서 Type으로 수정했습니다 ㅎㅎ

ksundong commented on src/main/java/net/jung/chess/pieces/Piece.java:
```diff
@@ -0,0 +1,48 @@
+public class Piece {
+    private final Color color;
+    private final Name name;
+    private final char representation;
```
name에 이미 있는데 representation을 다시 선언해주는 것은 중복입니다~

rla36 replied:
넵 Type(구 Name)에 접근해서 representation을 리턴하도록해야겠네요!!

ksundong commented on src/main/java/net/jung/chess/pieces/Piece.java:
```diff
@@ -0,0 +1,48 @@
+
+    public Name getName() { return name; }
+
+    public char getRepresentation() { return representation; }
```
여기서 name의 representation을 꺼내는 편이 더 좋아보이네요.

rla36 replied:
넵!!

ksundong commented on src/main/java/net/jung/chess/pieces/Piece.java:
```diff
@@ -0,0 +1,48 @@
+        return new Piece(color, name);
+    }
+
+    public static Piece createWhitePawn() { return new Piece(Color.WHITE, Name.PAWN); }
+    public static Piece createBlackPawn() { return new Piece(Color.BLACK, Name.PAWN); }
+    public static Piece createWhiteKnight() { return new Piece(Color.WHITE, Name.KNIGHT); }
+    public static Piece createBlackKnight() { return new Piece(Color.BLACK, Name.KNIGHT); }
+    public static Piece createWhiteRook() { return new Piece(Color.WHITE, Name.ROOK); }
+    public static Piece createBlackRook() { return new Piece(Color.BLACK, Name.ROOK); }
+    public static Piece createWhiteBishop() { return new Piece(Color.WHITE, Name.BISHOP); }
+    public static Piece createBlackBishop() { return new Piece(Color.BLACK, Name.BISHOP); }
+    public static Piece createWhiteQueen() { return new Piece(Color.WHITE, Name.QUEEN); }
+    public static Piece createBlackQueen() { return new Piece(Color.BLACK, Name.QUEEN); }
+    public static Piece createWhiteKing() { return new Piece(Color.WHITE, Name.KING); }
+    public static Piece createBlackKing() { return new Piece(Color.BLACK, Name.KING); }
```
사용되지 않는 메서드네요.
그리고 색까지 정의한 메서드를 선언하기 보다는, 색을 인자로 받아서 처리해주면 어떨까요?
그럼 장황하지 않은 느낌이 들 것 같습니다.

rla36 replied:
저도 미션 요구사항에 따라 구현한 메서드입니다..ㅜㅜ
상황에 따라 유연하게 적용할 수 없는 메서드라 미션에서 시키는대로 만들기만하고 사용하진 않은것같습니다..

rla36 replied:
디온님 말씀대로 색을 인자로 받아서 처리하면, 여러모로 중복제거도 가능할것같고 편리할 것 같습니다.!!

궁금한 점이, (아래아래 리뷰에서) Piece createPiece(Color color, Type type)을 제거하거나 private으로 전환하라고 하신 이유가, 매개변수로 Color 와 Type을 받아야 하니까 책임이 명확하지 않기 때문이라고 생각했습니다. 메서드  createWhitePawn 하나가 있다면 white Pawn만을 생성하는 역할만을 담당해 책임이 명확한반면, createPiece는 모든 종류의 흑백 말을 다 만들수있으니 책임이 명확하지 않으니까요.
그런데 여기서 createPawn(Color color)와 같이 매개변수로 색깔을 받게 되면, createPiece와 마찬가지로 책임이 불명확해지는 것이 아닌가 궁금합니다. 매개변수로 Color하나를 받으나, Color에 Type까지 두개 받으나 별 차이가 없다고 (식견 없는 제 눈에는) 보이는데, 왜 createPiece는 지워야하고 createPawn(Color color)은 괜찮은 것인가요..?
아니면 제가 createPiece를 지워야하는 이유를 잘못 짐작하고 있는 것인가요?
질문이 장황해서 죄송합니당..

ksundong replied:
유연한 것이 모두 좋은 것은 아닙니다. 비즈니스 요구사항에 맞춰 유연하게 대응 가능한 것이 좋은거죠.

---

createPiece를 지우라고 말씀드렸던 이유는 이를 만족하는 다른 메서드가 있음에도 사용하셔서 말씀 드렸던 거구요.
개인적으로 생각하기에는 유연한 메서드는 만들되, 이를 외부에는 감추고, 외부에는 비즈니스를 파악하기 쉬운 메서드를 제공하는 것이 좋다고 생각합니다.
좋은 질문 감사해요!

rla36 replied:
>  외부에는 비즈니스를 파악하기 쉬운 메서드를 제공하는 것이 좋다고 생각합니다. 

넵 이해 잘했어요 ㅎㅎ 알겠습니다!

ksundong commented on src/main/java/net/jung/chess/utils/StringUtils.java:
```diff
@@ -0,0 +1,11 @@
+
+    public static String appendNewLine(String s){
+        return s+NEWLINE;
+    }
```
`s` 같은 네이밍보단 `string`을 써줍시다.

ksundong commented on src/main/java/net/jung/chess/pieces/Piece.java:
```diff
@@ -0,0 +1,48 @@
+    public char getRepresentation() { return representation; }
+
+
+    public static Piece createPiece(Color color, Name name){
+        return new Piece(color, name);
+    }
```
이 메서드를 제거해주시거나 private으로 바꾸는 편이 좋을 것 같아요.

rla36 replied:
createWhitePawn()와 같이 색깔과 타입이 이미 정해져있는 팩토리메서드를 만들어줬기 때문에 필요가 없어진 것이죠??
그렇게 되면 외부 클래스에서 이 메서드를 사용해서 중복제거했던 것들도(Color을 변수로 받아 pawn제외 piece를 색깔별로 초기화하는 메서드 등...)  케이스별로 다 풀어써야하는데 createWhitePawn과 같은 팩토리메서드는 너무 유연성이 없다고 생각이 드네요ㅜㅜ 일단 알겠습니다.. 수정할게요..

ksundong commented on src/test/java/net/jung/chess/BoardTest.java:
```diff
@@ -1,73 +1,74 @@
-        String expectedLayout = "********\nPPPPPPPP\n********\n********\n********\n********\npppppppp\n********";
-        assertThat(board.print()).isEqualTo(expectedLayout);
+    @DisplayName("addWhitePawn(Piece pawn)의 매개변수로 blackPawn을 넘기면 whitePawnList에 더해지지 않고 반대도 마찬가지이다.")
+    void addBlackPawnToWhitePawnList() {
```
이건 엄연한 예외상황인데, 이를 무시할지, 예외를 발생시킬지는 개발자의 마음입니다.
이를 무시하는 선택을 해주셨군요. 👍

ksundong commented on src/test/java/net/jung/chess/BoardTest.java:
```diff
@@ -1,73 +1,74 @@
-        board.addWhitePawn(blackPawn);
-        board.addBlackPawn(whitePawn);
+    @DisplayName("initialize()는 32개 모든 말을 체스판 위에 올린다.")
+    void InitializeAddsEveryPiece(){
```
메서드는 camelCase로 작성해주세요~

rla36 replied:
넵넵

ksundong commented on src/test/java/net/jung/chess/BoardTest.java:
```diff
@@ -1,73 +1,74 @@
+    @DisplayName("initialize()는 32개 모든 말을 체스판 위에 올린다.")
+    void InitializeAddsEveryPiece(){
+        board.initialize();
+        String blankRank = appendNewLine("********");
```
같은 선언을 다른 곳에서도 봤던 것 같은데, 추출시켜보는 것도 생각해보세요!

rla36 replied:
Board클래스에
public final String BLANK_RANK = appendNewLine("********");
을 추가해 boardLayoutToString메서드와 테스트 메서드에서 쓰도록 수정했습니다

ksundong commented on src/test/java/net/jung/chess/BoardTest.java:
```diff
@@ -1,73 +1,74 @@
-                () -> assertThat(board.blackPawnListSize()).isEqualTo(0)
-                );
-
+                () -> assertThat(board.size()).isEqualTo(32),
```
32를 여기 써주는 것도 좋지만, 상수로 빼는 편도 좋을 것 같습니다.

ksundong commented on src/test/java/net/jung/chess/BoardTest.java:
```diff
@@ -1,73 +1,74 @@
+                    +appendNewLine("PPPPPPPP")
+                    +blankRank+blankRank+blankRank+blankRank
+                    +appendNewLine("pppppppp")
+                    +appendNewLine("rnbqkbnr"))
```
저는 blankRank 또한 한 줄 씩 입력해주는게 좋은 것 같습니다.

ksundong commented on src/test/java/net/jung/chess/pieces/PawnTest.java:
```diff
@@ -1,11 +1,11 @@
-import static org.junit.jupiter.api.Assertions.assertEquals;
 
 
 public class PawnTest {
```
`Pawn` 클래스가 없어졌다면 테스트도 없애주는 편이 좋을 것 같아요.

ksundong commented on src/test/java/net/jung/chess/pieces/PieceTest.java:
```diff
@@ -0,0 +1,53 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+public class PieceTest {
```
JUnit5에서는 테스트 클래스에 `public` 접근 제어자는 붙여주지 않는 것이 일반적입니다!

ksundong commented on src/test/java/net/jung/chess/pieces/PieceTest.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    private void verifyPiece(final Piece piece, final Color color, final Name name, final char representation){
+        assertThat(piece.getColor()).isEqualTo(color);
+        assertThat(piece.getName()).isEqualTo(name);
+        assertThat(piece.getRepresentation()).isEqualTo(representation);
```
여기는 `assertAll()`을 사용하지 않으신 이유가 있나요?

rla36 replied:
앗 아뇨... 쓰는 것이 좋을것같은데 뺴먹었습니다 수정했습니다!!

ksundong commented on src/test/java/net/jung/chess/pieces/PieceTest.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Test
+    @DisplayName("isBlack isWhite는 색깔이 일치하면 true 아니면 false 리턴한다.")
+    void isColorCheck(){
```
위의 메서드랑 네이밍 순서가 다르죠? 주의해주세요~

ksundong reviewed (changes requested):
안녕하세요! Jung 리뷰어 디온입니다. 4단계 까지 빨리오셨네요! 축하드립니다. 🎉

값 객체에 대해서 제대로 공부를 하셨네요. 💯 
맞습니다. 값 객체는 우리가 얘기하는 리터럴 처럼 값 그 자체여야 합니다.
따라서 값 객체로 사용할 때에는 `equals()`&`hashCode()` 오버라이딩을 해주는 것이 맞습니다.

다만 명시적으로 요구사항에서 위의 내용을 요구하지는 않아서 다른 분들의 코드에 리뷰하지 않았습니다.
그리고 private 생성자를 만드는 부분도 VO에 필수적인 내용은 아니지만, 요구 사항에 있어서 리뷰했어요.

- 'VO는 인스턴스 생성 후 상태를 변경할 수 없다.(불변 객체여야 한다)'는 필요조건이고, 상태를 변경할 수 없다고 모두 VO는 아닙니다.
- 'VO는 `equals()`와 `hashCode()`를 재정의 해야한다. 역시 필요조건이고, 이를 재정의 한다고 모두 VO는 아닙니다.
- private 생성자는 필요조건도, 충분조건도 아닙니다.

하지만 현재 요구사항에서 굳이 VO를 만들어줄 필요는 없다고 생각하여 리뷰하지 않았습니다.

## 리뷰 내용

- 현재 시점에서 필요없는 코드는 과감히 빼주세요.
- 반복되는 부분은 추출을 통해서 반복을 제거해주세요.
- 메서드 네이밍 컨벤션을 지켜주세요.
- 일관적인 코드 스타일을 지켜주세요.
- 좀 더 좋은 네이밍을 고민해주세요.
- 테스트로만 검증하고, 정작 실제 코드에서는 다른 코드를 사용하는 것이 자주 발견됩니다. 신경써주세요!

미션 수행하시느라 정말 고생 많으셨고, 자잘한 피드백이 조금 많습니다. 😅
궁금하시거나, 이해가 잘 안되시는 부분들은 코멘트 꼭 달아주시고, 피드백 반영하신 후에는 다시 리뷰요청해주세요!

tmdgusya commented on src/main/java/net/jung/chess/utils/StringUtils.java:
```diff
@@ -0,0 +1,11 @@
+
+    private StringUtils(){    }
+
+    public static String appendNewLine(String s){
```
오 저는 Builder 로 이을 생각만했는데, 이렇게 String 에 관한 처리만 해서 return 해주는 방식이 조금더 간결할 수 있겠다 라는 생각이 드네요.

rla36 replied:
반갑슴다 로치ㅎㅎ 코멘트 감사감사해용

tmdgusya commented on src/test/java/net/jung/chess/BoardTest.java:
```diff
@@ -44,18 +45,29 @@ void createBoardAddPawns() throws Exception{
+                .append("********").append(System.getProperty("line.separator"))
+                .append("********").append(System.getProperty("line.separator"))
+                .append("********").append(System.getProperty("line.separator"))
+                .append("pppppppp").append(System.getProperty("line.separator"))
```
```java
public static final String NEWLINE 
```
해당 부분이 StringUtils 에 Public 으로 열려있는 만큼 System.getProperty("line.separator") 보다는 
NewLine 을 쓰는것이 좋아보이기도 합니다.

tmdgusya commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -40,45 +50,67 @@ public Piece findBlackPiece(int index) {
         for (int i = 0; i < colorPawnSize; i++) {
             pawnReps.append(pieceList.get(i).getRepresentation());
+            if(i==7)
+                pawnReps.append(System.getProperty("line.separator"));
```
```java
System.getProperty("line.separator")
```
해당 부분도 StringUtils 에 열려있는 NEWLINE 을 쓰는것이 바람직해 보입니다.

tmdgusya commented on src/main/java/net/jung/chess/utils/StringUtils.java:
```diff
@@ -1,7 +1,7 @@
 
 public class StringUtils {
-    public static final String NEWLINE = System.getProperty("line.separator");
+    private static final String NEWLINE = System.getProperty("line.separator");
```
해당 부분을 private 으로 바꾸셨네요.
지금 코드를 봤을때 외부에서 System.getProperty("line.separator"); 을 쓰는곳이 많은데
차라리 public 으로 열고 외부에서 static import 한다음에 `NEWLINE` 으로 간단하게 사용하는건 어떨까요?

rla36 committed:
Name에서 Type으로 enum 수정 및 pieces로 Color, Type 이동

Color, Type는 pieces의 속성이므로 pieces 디렉토리 아래 이동했습니다.

rla36 committed:
Board클래스 내 리네이밍 및 리팩토링

rla36 committed:
Piece내 representation 제거, getRepresentation 리팩토링

Representation은 Type에 이미 선언되어 있으므로 Piece내에서 다시
인스턴스변수로 두는 것은 중복입니다. 따라서 변수를 제거하고
getRepresentation메서드에서는 Type에 접근해 representation을 리턴하도록
했습니다.

rla36 committed:
Test클래스 리네이밍 및 리팩토링

Board와 BoardTest에서 공통적으로 쓰이는 string을 Board에 public final
String으로 선언해서 양쪽에서 쓸 수 있도록했습니다.

rla36 committed:
createPiece제거와 이에  수반한  리팩토링

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,122 @@
-    private List<Pawn> whitePawnList = new ArrayList();
+    private List<Piece> blackPieceList = new ArrayList();
+    private List<Piece> whitePieceList = new ArrayList();
+    public final String BLANK_RANK = appendNewLine("********");
```
요건 `static`이 더 좋을 것 같네요.

rla36 replied:
모든 인스턴스에서 BLANK_RANK는 동일하게 쓰이니까 static이 좋을것같네요 
그 생각을 미처 못했는데 리뷰 감사합니다!

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,122 @@
-        for (int i = 0; i < MAX_PAWNS_SIZE; i++) {
-            addWhitePawn(new Pawn(Color.WHITE));
-            addBlackPawn(new Pawn(Color.BLACK));
+        addBlackPiece(Piece.createBlackRook());
+        addBlackPiece(Piece.createBlackKnight());
+        addBlackPiece(Piece.createBlackBishop());
+        addBlackPiece(Piece.createBlackQueen());
+        addBlackPiece(Piece.createBlackKing());
+        addBlackPiece(Piece.createBlackBishop());
+        addBlackPiece(Piece.createBlackKnight());
+        addBlackPiece(Piece.createBlackRook());
```
`initializeBlackPieces` 정도로 추출해도 괜찮을 것 같아요.

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,122 @@
+
+    public void initializePawns(Color color) {
+        int maxPawnsSize = 8;
+        if(color == Color.WHITE) {
+        for (int i = 0; i < maxPawnsSize; i++) {
+            addWhitePiece(Piece.createWhitePawn());
+             }
+        return;
+        }
```
코드 정렬이 안되어 있네요.

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -1,85 +1,122 @@
+        StringBuilder piecesRepresentation = new StringBuilder();
+        for (int i = 0, colorPawnSize = pieceList.size(); i < colorPawnSize; i++) {
+            piecesRepresentation.append(pieceList.get(i).getRepresentation());
+            if(i==7)
+                piecesRepresentation.append(System.getProperty("line.separator"));
```
모든 if 문에 블록을 만들어주세요.

ksundong commented on src/main/resources/logback.xml:
```diff
@@ -10,7 +10,7 @@
 
     <!-- 자바 패키지별 log level을 설정한다. -->
-    <logger name="net.honux.chess" level="debug" />
+    <logger type="net.honux.chess" level="debug" />
```
이 파일이 바뀐 이유는 뭘까요?

rla36 replied:
저는 바꾼 적 없어서 잘 모르겠습니다...! enum 이름을 Name에서 Type으로 바꾸는 과정에서 자동 수정된 것이 아닐까 짐작해봅니다..

ksundong replied:
이런 부분도 커밋하실 때 꼭 체크해주세요.
일단 원래대로 돌려주세요~

rla36 replied:
미션 5 진행하면서 수정했습니다!!

ksundong commented on src/test/java/net/jung/chess/BoardTest.java:
```diff
@@ -1,73 +1,77 @@
                 () -> {
-                    board.addWhitePawn(whitePawn);
+                    board.addWhitePiece(whitePawn);
                     assertAll("흰색 폰을 보드에 올린 후 size, findPawn메서드 검증한다.",
                             () -> assertThat(board.size()).isEqualTo(1),
-                            () -> assertThat(board.findWhitePawn(0)).isEqualTo(whitePawn));
+                            () -> assertThat(board.findWhitePiece(0)).isEqualTo(whitePawn));
                 },
                 () -> {
-                    board.addBlackPawn(blackPawn);
+                    board.addBlackPiece(blackPawn);
                     assertAll("검은색 폰을 보드에 올린 후 size, findPawn메서드 검증한다.",
                             () -> assertThat(board.size()).isEqualTo(2),
-                            () -> assertThat(board.findBlackPawn(0)).isEqualTo(blackPawn));
+                            () -> assertThat(board.findBlackPiece(0)).isEqualTo(blackPawn));
                 }
         );
```
저는 테스트 하나에 assertion 하나를 추천합니다.
이 테스트는 assertAll이라는 단위 두 개가 묶여있네요

rla36 replied:
넵 미션2 때 작성했던 것 같은데, 보드 위에 피스 두개를 올린 후 size()와 findWhitePiece(int index), findBlackPiece(int index) 를 검증하기 위해 저렇게 했습니다. assertAll 두개 사이에 addBlackPiece(blackPawn)만 없으면 하나로 통합할수 있을 것 같은데, 보드에 피스를 한 개 초과해서 올린 후 메서드를 검증하고 싶었던 것이라 assertAll두개를 assertAll로 묶는 방식으로 타협했습니다.. 다음부터 테스트 메서드에는 assertion 하나만 쓰는 방식으로 최대한 해보겠습니다!

ksundong replied:
그런 의도가 있었다면 살짝 아쉽지만 그대로 진행해보시는 것도 좋을 것 같습니다.

ksundong reviewed (changes requested):
좋은 고민을 많이 해주신 것 같습니다.
당장 다음 단계를 진행해도 괜찮을 것 같지만, 좀만 더 생각해보세요.
그리고 커밋하실 때 파일을 잘 확인하고 올려주세요!
궁금하신점은 코멘트 또 달아주세요~

rla36 committed:
Board클래스 수정. 접근제어자 변경, 정렬, 리뷰 반영

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -11,7 +11,7 @@
     private List<Piece> blackPieceList = new ArrayList();
     private List<Piece> whitePieceList = new ArrayList();
-    public final String BLANK_RANK = appendNewLine("********");
+    public static final String BLANK_RANK = appendNewLine("********");
```
이 변수의 위치도 자바 컨벤션이 있습니다.
맨 위로 올려주시면 됩니다~

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -27,7 +27,7 @@ public void addBlackPiece(Piece piece) {
 
     public int size() {
-        return whitePieceList.size() + blackPieceList.size();
+        return whitePieceListSize() + blackPieceListSize();
```
👍

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -47,6 +47,28 @@ public Piece findBlackPiece(int index) {
+    public void initializePawns(Color color) {
+        int maxPawnsSize = 8;
+        if(color == Color.WHITE) {
+            for (int i = 0; i < maxPawnsSize; i++) {
+            addWhitePiece(Piece.createWhitePawn());
+            }
+        return;
```
들여쓰기를 신경써주세요~
early return은 좋습니다.

ksundong commented on src/main/java/net/jung/chess/Board.java:
```diff
@@ -93,10 +101,11 @@ public String getBlackPiecesRepresentation() {
+        for (int i = 0; i < pieceList.size(); i++) {
             piecesRepresentation.append(pieceList.get(i).getRepresentation());
-            if(i==7)
+            if(i==7) {
```
코드 커밋 전에 정렬 꼭 단축키 이용해서 해주세요~

ksundong reviewed (approved):
Jung 고생많으셨습니다. 좀 더 다듬어주셨으면 좋겠다고 생각되는 부분에 코멘트 남겼습니다.
다음단계 진행해주세요~
